task/control-plane-agent/src/mgs_common.rs

```diff
@@ -143,14 +153,153 @@ impl MgsCommon {
             CabooseError::NoImageHeader => panic!(),
         })
     }
+
+    /// If the targeted component is the SP_ITSELF, then having reset itself,
+    /// it will not be able to respond to the later reset_trigger message.
+    ///
+    /// So, after getting an ACK for the prepare message, MGS will send and
+    /// retry the reset_trigger message until it gets rejected for lack of
+    /// a corresponding prepare message.
+    ///
+    /// If the targeted component is not the SP_ITSELF, it may still have impact
+    /// on the SP if reset, either now or in a future implementation.
+    /// However, for some components, the SP will be able to send an
+    /// acknowledgement and retrying the trigger message will not be effective.
+    /// The implementation in the control plane should handle both cases.
+    pub(crate) fn reset_component_prepare(
+        &mut self,
+        component: SpComponent,
+    ) -> Result<(), SpError> {
+        self.reset_component_requested = component;
+        Ok(())
+    }
+
+    /// ResetComponent is used in the context of the management plane
+    /// driving a firmware update.
+    ///
+    /// When an update is complete, or perhaps for handling update errors,
+    /// the management plane will need to reset a component or change
+    /// boot image selection policy and reset that component.
+    ///
+    /// The target of the operation is the management plane's SpComponent
+    /// and firmware slot.
+    /// For the RoT, that is SpComponent::ROT and slot 0(ImageA) or 1(ImageB)
+    /// or SpComponent::STATE0 and slot 0.
+    pub(crate) fn reset_component_trigger(
+        &mut self,
+        update: &SpUpdate,
+        component: SpComponent,
+    ) -> Result<(), SpError> {
+        if self.reset_component_requested != component {
+            return Err(SpError::ResetComponentTriggerWithoutPrepare);
+        }
+        // If we are not resetting the SP_ITSELF, then we may come back here
+        // to reset something else or to run another prepare/trigger on
+        // the same component.
+        self.reset_component_requested.id.fill(0);
+
+        // Resetting the SP through reset_component() is
+        // the same as through reset() until transient bank selection is
+        // figured out for the SP.
+        match component {
+            SpComponent::SP_ITSELF => {
+                task_jefe_api::Jefe::from(crate::JEFE.get_task_id())
+                    .request_reset();
+                // If `request_reset()` returns,
+                // something has gone very wrong.
+                panic!();
+            }
+            SpComponent::ROT => {
+                // mgs_{gimlet,psc,sidecar}.rs deal with any board specific
```

Comment:
This comment is confusing; it should instead live next to the `return Err(SpError...`) line.

I'd also suggest moving `match update.sprot_task().reset() { ...` into this block of the case statement; otherwise, you have to notice that all the other cases diverge and only this one falls through.

Reply:
Both suggestions fixed in https://github.com/oxidecomputer/hubris/pull/1261/commits/c67a55f3bd72ec7f01635da43ba0e53555405e7f